examples/nlp/nl_image_search.py

```diff
@@ -0,0 +1,612 @@
+"""
+
+projection_dims = 256
+num_projection_layers = 1
+dropout_rate = 0.1
+learning_rate = 0.003
+weight_decay = 0.0001
+num_epochs = 30
+batch_size = 256
+temperature = 0.07
+
+vision_encoder = create_vision_encoder(
+    num_projection_layers, projection_dims, dropout_rate
+)
+vision_encoder.summary()
+
+text_encoder = create_text_encoder(num_projection_layers, projection_dims, dropout_rate)
+text_encoder.summary()
+
+optimizer = tfa.optimizers.AdamW(learning_rate=learning_rate, weight_decay=weight_decay)
+loss = ContrastiveSimilarityLoss(temperature)
+dual_encoder = create_dual_encoder(text_encoder, vision_encoder)
+dual_encoder.compile(optimizer=optimizer, loss=loss)
+dual_encoder.summary()
+
+"""
+Note that training the model with 60,000 image-caption pairs, with a batch size of 256,
+takes around 11 minutes per epoch using a V100 GPU accelerator"""
+
+print(f"Number of examples (caption-image pairs): {train_example_count}")
+print(f"Batch size: {batch_size}")
+print(f"Steps per epoch: {train_example_count // batch_size}")
+train_data = get_dataset(os.path.join(tfrecords_dir, "train-*.tfrecord"), batch_size)
+history = dual_encoder.fit(train_data, epochs=num_epochs)
+vision_encoder.save("vision_encoder")
+text_encoder.save("text_encoder")
+
+"""
+Plotting the training loss:
+"""
+
+plt.plot(history.history["loss"])
+plt.xlabel("Epochs")
+plt.ylabel("Loss")
+plt.show()
+
+"""
+## Search for images using natural language queries
+
+We can then retrieve images corresponding to natural language queries via
+the following steps:
+
+1. Generate embeddings for the images by feeding them into the `vision_encoder`.
+2. Feed the natural language query to the `text_encoder` to generate a query embedding.
+3. Compute the similarity between the query embedding and the image embeddings
+in the index to retrieve the indices of the top matches.
+4. Look up the paths of the top matching images to display them.
+
+Note that, after training the `dual encoder`, the only the fine-tuned `vision_encoder`
+and `text_encoder` models will be used, while the `dual_encoder` model will be discarded.
+"""
+
+"""
+### Generate embeddings for the images
+
+We load the images and feed them into the `vision_encoder` to generate their embeddings.
+In large scale systems, this step is performed using a parallel data processing frameworks,
+such as [Apache Spark](https://spark.apache.org) or [Apache Beam](https://beam.apache.org).
+Generating the image embeddings will take a few minutes.
+"""
+
+vision_encoder = tf.saved_model.load("vision_encoder")
+text_encoder = tf.saved_model.load("text_encoder")
+
+image_embeddings = []
+counter = 0
+
+
+def read_image(image_path):
+    raw_image = tf.io.read_file(image_path)
+    image_array = tf.image.decode_jpeg(raw_image, channels=3)
+    return tf.image.resize(image_array, (299, 299))
+
+
+print(f"Image embeddings generation started...")
+print(f"Generating embeddings for {len(image_paths)} images...")
+for image_batch in (
```

Comment:
This paragraph can be replaced with

`image_embeddings  = vision_encoder.predict(dataset)`

no?

Reply:
Nice one! Thanks